piet/src/gradient.rs

```diff
@@ -84,13 +84,56 @@ pub trait GradientStops {
 ///
 /// [`UnitPoint`]: struct.UnitPoint.html
 /// [`FixedGradient`]: struct.FixedGradient.html
+#[derive(Debug, Clone)]
 pub struct LinearGradient {
     start: UnitPoint,
     end: UnitPoint,
     stops: Vec<GradientStop>,
 }
 
+/// A description of a radial gradient in the unit rect, which can be resolved
+/// to a fixed gradient.
+///
+/// The center is given in `UnitPoint` coordinates.
+///
+/// The `center` and `radius` describe a sphere. `origin` describes the angle
+/// of the gradient on that sphere, useful for simulating lighting effects.
+/// see [configuring a radial gradient][config] for a fuller explanation.
+///
+/// By default, `origin` is equal to `center`. This can be changed during construction
+/// with the [`with_origin`] builder method.
+///
+/// The [`ScaleMode`] describes how the gradient is mapped to a non-square
+/// rectangle; by default this will expand on the longest axis, but this can
+/// be changed with the [`with_scale_mode`] builder method.
+///
+/// [config]: https://docs.microsoft.com/en-us/windows/win32/direct2d/direct2d-brushes-overview#configuring-a-radial-gradient
+/// [`ScaleMode`]: enum.ScaleMode.html
+/// [`with_origin`]: struct.RadialGradient.html#method.with_origin
+/// [`with_scale_mode`]: struct.RadialGradient.html#method.with_scale_mode
+#[derive(Debug, Clone)]
+pub struct RadialGradient {
+    center: UnitPoint,
+    origin: UnitPoint,
+    radius: f64,
+    stops: Vec<GradientStop>,
+    scale_mode: ScaleMode,
+}
+
+/// Mappings from the unit square into a non-square rectangle.
+#[derive(Debug, Clone)]
+pub enum ScaleMode {
+    /// The unit 1.0 is mapped to the smaller of width & height. All
```

Comment:
I think talking about "clipping" here might be confusing, as there's no additional clipping logic based on the rect. You might well see behavior that could be described as clipping when the radius is 0.5.